File: examples/python/seaweedfs/seaweedfs_fuse.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import os
import time

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.seaweedfs import SeaweedFSConfig, SeaweedFSResource

load_dotenv(".env.development")

config = SeaweedFSConfig(
    bucket=os.environ.get("SEAWEEDFS_BUCKET", "mirage-demo"),
    endpoint_url=os.environ.get("SEAWEEDFS_ENDPOINT", "http://localhost:8333"),
    access_key_id=os.environ.get("SEAWEEDFS_ACCESS_KEY", "any"),
    secret_access_key=os.environ.get("SEAWEEDFS_SECRET_KEY", "any"),
)

resource = SeaweedFSResource(config)

SEED_KEYS = ("/seaweedfs/data/example.jsonl", "/seaweedfs/data/config.json",
             "/seaweedfs/notes.txt")


async def seed(ws: Workspace) -> None:
    await ws.ops.write(
        "/seaweedfs/data/example.jsonl",
        b'{"event":"queue-operation","tool":"mirage"}\n'
        b'{"event":"read","tool":"mirage"}\n'
        b'{"event":"queue-operation","tool":"other"}\n')
    await ws.ops.write(
        "/seaweedfs/data/config.json",
        b'{"name":"mirage","version":1,"tags":["s3","seaweedfs"]}')
    await ws.ops.write("/seaweedfs/notes.txt", b"hello from seaweedfs\n")


async def cleanup(ws: Workspace) -> None:
    for key in SEED_KEYS:
        await ws.execute(f"rm {key}")


with Workspace(
    {"/seaweedfs/": resource},
        mode=MountMode.WRITE,
        fuse=True,
) as ws:
    asyncio.run(seed(ws))
    time.sleep(1)
    mp = ws.fuse_mountpoint

    print(f"=== FUSE MODE: SeaweedFS mounted at {mp} ===\n")

    print("--- os.listdir() ---")
    for e in os.listdir(f"{mp}/seaweedfs/data"):
        print(f"  {e}")

    print("\n--- open() + read ---")
    with open(f"{mp}/seaweedfs/data/example.jsonl") as f:
        for i, line in enumerate(f):
            if i >= 3:
                break
            print(f"  [{i}] {line.strip()[:100]}...")

    print("\n--- os.path.getsize() ---")
    size = os.path.getsize(f"{mp}/seaweedfs/data/example.jsonl")
    print(f"  size: {size} bytes")

    print(f"\n>>> FUSE mounted at: {mp}")
    print(">>> Open another terminal and run:")
    print(f">>>   ls {mp}/seaweedfs/data/")
    print(f">>>   cat {mp}/seaweedfs/data/config.json")
    print(">>> Press Enter to unmount and exit...")
    try:
        input()
    except EOFError:
        pass

    asyncio.run(cleanup(ws))

    records = ws.ops.records
    total = sum(r.bytes for r in records)
    print(f"\nStats: {len(records)} ops, {total} bytes transferred")

File: examples/typescript/seaweedfs/seaweedfs_fuse.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

// SeaweedFS mounted as a real FUSE filesystem. Seeds a few objects, then
// external processes can `cat $mp/seaweedfs/data/config.json` just like a
// local directory.
//
// Requires: macFUSE / libfuse3 + @zkochan/fuse-native (see /typescript/setup/fuse).
// Point it at a SeaweedFS S3 gateway (default http://localhost:8333).
// Loads credentials from .env.development at the repo root.
import {
  FuseManager,
  MountMode,
  SeaweedFSResource,
  Workspace,
  type SeaweedFSConfig,
} from '@struktoai/mirage-node'
import dotenv from 'dotenv'
import { resolve, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const __HERE = dirname(fileURLToPath(import.meta.url))
dotenv.config({ path: resolve(__HERE, '../../../.env.development') })

const SEED_KEYS = [
  '/seaweedfs/data/example.jsonl',
  '/seaweedfs/data/config.json',
  '/seaweedfs/notes.txt',
]

function configFromEnv(): SeaweedFSConfig {
  return {
    bucket: process.env.SEAWEEDFS_BUCKET ?? 'mirage-demo',
    endpoint: process.env.SEAWEEDFS_ENDPOINT ?? 'http://localhost:8333',
    accessKeyId: process.env.SEAWEEDFS_ACCESS_KEY ?? 'any',
    secretAccessKey: process.env.SEAWEEDFS_SECRET_KEY ?? 'any',
  }
}

async function seed(ws: Workspace): Promise<void> {
  await ws.execute(
    `echo '{"event":"queue-operation","tool":"mirage"}' > /seaweedfs/data/example.jsonl`,
  )
  await ws.execute(`echo '{"event":"read","tool":"mirage"}' >> /seaweedfs/data/example.jsonl`)
  await ws.execute(
    `echo '{"event":"queue-operation","tool":"other"}' >> /seaweedfs/data/example.jsonl`,
  )
  await ws.execute(
    `echo '{"name":"mirage","version":1,"tags":["s3","seaweedfs"]}' > /seaweedfs/data/config.json`,
  )
  await ws.execute('echo "hello from seaweedfs" > /seaweedfs/notes.txt')
}

async function main(): Promise<void> {
  const cfg = configFromEnv()
  const ws = new Workspace(
    { '/seaweedfs/': new SeaweedFSResource(cfg) },
    { mode: MountMode.WRITE },
  )

  try {
    await seed(ws)

    const fm = new FuseManager()
    const mp = await fm.setup(ws)
    let cleaned = false
    const handler = (sig: NodeJS.Signals): void => {
      if (cleaned) return
      cleaned = true
      void (async (): Promise<void> => {
        try {
          for (const key of SEED_KEYS) await ws.execute(`rm ${key}`)
        } catch {}
        try {
          await fm.close(ws)
        } catch {}
        try {
          await ws.close()
        } catch {}
        console.error(`\n>>> unmounted ${mp}`)
        process.exit(sig === 'SIGINT' ? 130 : 143)
      })()
    }
    process.on('SIGINT', handler)
    process.on('SIGTERM', handler)

    console.log(`=== FUSE MODE — SeaweedFS at ${cfg.endpoint} (bucket ${cfg.bucket}) ===`)
    console.log(`  mountpoint = ${mp}\n`)

    try {
      console.log('--- virtual executor: stats via /seaweedfs ---')
      const ls = await ws.execute('ls /seaweedfs/data')
      console.log(`  ls /seaweedfs/data : ${ls.stdoutText.trim().split('\n').join(', ')}`)
      const stat = await ws.execute('stat /seaweedfs/data/example.jsonl')
      console.log(`  stat example.jsonl : ${stat.stdoutText.trim()}`)
      const grep = await ws.execute('grep -c queue-operation /seaweedfs/data/example.jsonl')
      console.log(`  grep -c            : ${grep.stdoutText.trim()}`)

      console.log()
      console.log('>>> Mount is live. From ANOTHER terminal you can:')
      console.log(`>>>   ls  ${mp}/seaweedfs/data/`)
      console.log(`>>>   cat ${mp}/seaweedfs/data/config.json`)
      console.log(`>>>   wc -l ${mp}/seaweedfs/data/example.jsonl`)
    } finally {
      for (const key of SEED_KEYS) await ws.execute(`rm ${key}`)
      await fm.close(ws)
      console.log(`\nafter unmount: ws.fuseMountpoint = ${ws.fuseMountpoint ?? 'null'}`)
    }
  } finally {
    await ws.close()
  }
}

main().catch((err: unknown) => {
  console.error(err)
  process.exit(1)
})
